src/builder.ts

```diff
@@ -95,56 +106,58 @@ export default class Builder {
 							if (data == null) {
 								return;
 							}
-							if (data.error) {
-								errored = true;
-								dup.destroy(new Error(data.error));
-							} else {
-								// Store image layers, so that they can be deleted by the caller
-								// if necessary
-								const sha = Utils.extractLayer(data.stream);
-								if (sha !== undefined) {
-									self.layers.push(sha);
+							try {
+								if (data.error) {
+									throw new Error(data.error);
+								} else {
+									// Store image layers, so that they can be
+									// deleted by the caller if necessary
+									const sha = Utils.extractLayer(data.stream);
+									if (sha !== undefined) {
+										self.layers.push(sha);
+									}
+									this.emit('data', data.stream);
 								}
-
-								this.emit('data', data.stream);
+							} catch (error) {
+								daemonStream.unpipe();
+								reject(error);
 							}
 						}),
 					);
 
-				// Catch any errors the stream produces
-				outputStream.on('error', (err: Error) => {
-					errored = true;
-					self.callHook(hooks, 'buildFailure', handler, err, self.layers);
-				});
-				dup.on('error', (err: Error) => {
-					errored = true;
-					self.callHook(hooks, 'buildFailure', handler, err, self.layers);
-				});
-
-				// Setup the buildSuccess hook. This handler is not called on
-				// error so we can use it to propagate the success information
-				outputStream.on('end', () => {
-					if (!errored) {
-						this.callHook(
-							hooks,
-							'buildSuccess',
-							handler,
-							_.last(this.layers),
-							this.layers,
-						);
-					}
+				outputStream.on('error', (error: Error) => {
+					daemonStream.unpipe();
+					reject(error);
 				});
+				outputStream.on('end', () =>
+					// The 'end' event was observed to be emitted under error
+					// conditions, hence the test for streamError.
+					streamError ? reject(streamError) : resolve(),
+				);
 				// Connect the output of the docker daemon to the duplex stream
 				dup.setReadable(outputStream);
-			})
-			.catch((err: Error) => {
-				// Call the plugin's error handler
-				self.callHook(hooks, 'buildFailure', handler, err, self.layers);
 			});
+		});
```

Comment:
There is no `.catch` here, but see `Bluebird.join()` below that has a `.catch`.

Reply:
Maybe a comment detailing that we want errors to bubble up here?

Reply:
Added